laterpay/application/Controller/Setting.php

```diff
@@ -604,35 +662,79 @@ public function get_checkbox_field_markup( $field = null ) {
+        $categories = isset( $field['categories'] ) ? $field['categories'] : array();
+        $unlimited  = get_option( 'laterpay_unlimited_access' ) ? get_option( 'laterpay_unlimited_access' ) : array();
+
+        $inputs_markup  = '';
+        $count          = 1;
+
+        if ( $role ) {
+            foreach ( $categories as $id => $name ) {
+                $need_default   = ! isset( $unlimited[$role] ) || ! $unlimited[$role];
+                $is_none_or_all = in_array( $id, array( 'none', 'all' ), true );
+                $is_selected    = ! $need_default ? in_array( $id, $unlimited[$role] ) : false;
+
+                $inputs_markup .= '<input type="checkbox" ';
+                $inputs_markup .= 'id="lp_category--' . $role . $count . '"';
+                $inputs_markup .= 'class="lp_category-access-input';
+                $inputs_markup .= $is_none_or_all ? ' lp_global-access" ' : '" ';
+                $inputs_markup .= 'name="laterpay_unlimited_access[' . $role . '][]"';
+                $inputs_markup .= 'value="' . $id . '" ';
+                $inputs_markup .= $is_selected || ( $need_default && $id === 'none' ) ? 'checked' : '';
+                $inputs_markup .= '>';
+                $inputs_markup .= '<label class="lp_category-access-label';
+                $inputs_markup .= $is_none_or_all ? ' lp_global-access" ' : '" ';
+                $inputs_markup .= 'for="lp_category--' . $role . $count . '">';
+                $inputs_markup .= $name;
+                $inputs_markup .= '</label>';
 
-        $default_roles    = array( 'administrator', 'editor', 'contributor', 'author', 'subscriber' );
-        $has_custom_roles = false;
-        $option_value     = get_option( 'laterpay_unlimited_access_to_paid_content' );
+                $count += 1;
+            }
+        }
 
-        $inputs_markup = '';
-        foreach ( $wp_roles->roles as $role => $role_data ) {
-            if ( ! in_array( $role, $default_roles ) ) {
-                $has_custom_roles = true;
-                $inputs_markup .= '<label title="' . $role_data['name'] . '">';
-                $inputs_markup .= '<input type="checkbox" name="laterpay_unlimited_access_to_paid_content" value="' . $role . '" ';
-                if ( in_array( $role, ( array ) $option_value ) ) {
-                    $inputs_markup .= 'checked';
+        echo $inputs_markup;
+    }
+
+    /**
+     * Validate unlimited access inputs before saving.
+     *
+     * @param $input
+     *
+     * return $valid array of valid values
+     */
+    public function validate_unlimited_access( $input ) {
+        $valid = array();
+
+        foreach ( $input as $role => $data ) {
+            if ( in_array( 'all', $data ) && in_array( 'none', $data ) && count( $data ) == 2 ) {
+                // unset option 'all', if option 'all' and option 'none' are selected at the same time
+                unset( $data[array_search( 'all', $data )] );
+            } elseif ( count( $data ) > 1 ) {
```

Comment:
@AliaksandrVahura-ScienceSoft I have modified the validation rules. Can you please review, if everything is correct like this?